fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Utils.java

```diff
@@ -536,4 +557,156 @@ private static <T extends ScalarOperator> void collect(ScalarOperator root, Clas
 
         root.getChildren().forEach(child -> collect(child, clazz, output));
     }
+
+    public static Set<MaterializedView> getRelatedMvs(int maxLevel, List<Table> tablesToCheck) {
+        Set<MaterializedView> mvs = Sets.newHashSet();
+        getRelatedMvs(maxLevel, 0, tablesToCheck, mvs);
+        return mvs;
+    }
+
+    public static void getRelatedMvs(int maxLevel, int currentLevel, List<Table> tablesToCheck, Set<MaterializedView> mvs) {
+        if (currentLevel >= maxLevel) {
+            return;
+        }
+        Set<MvId> newMvIds = Sets.newHashSet();
+        for (Table table : tablesToCheck) {
+            Set<MvId> mvIds = table.getRelatedMaterializedViews();
+            if (mvIds != null && !mvIds.isEmpty()) {
+                newMvIds.addAll(mvIds);
+            }
+        }
+        if (newMvIds.isEmpty()) {
+            return;
+        }
+        List<Table> newMvs = Lists.newArrayList();
+        for (MvId mvId : newMvIds) {
+            Database db = GlobalStateMgr.getCurrentState().getDb(mvId.getDbId());
+            if (db == null) {
+                continue;
+            }
+            Table table = db.getTable(mvId.getId());
+            if (table == null) {
+                continue;
+            }
+            newMvs.add(table);
+            mvs.add((MaterializedView) table);
+        }
+        getRelatedMvs(maxLevel, currentLevel + 1, newMvs, mvs);
+    }
+
+    // get all ref tables within and below root
+    public static List<Table> getAllTables(OptExpression root) {
+        List<Table> tables = Lists.newArrayList();
+        getAllTables(root, tables);
+        return tables;
+    }
+
+    private static void getAllTables(OptExpression root, List<Table> tables) {
+        if (root.getOp() instanceof LogicalScanOperator) {
+            LogicalScanOperator scanOperator = (LogicalScanOperator) root.getOp();
+            tables.add(scanOperator.getTable());
+        } else {
+            for (OptExpression child : root.getInputs()) {
+                getAllTables(child, tables);
+            }
+        }
+    }
+
+    public static boolean isValidMVPlan(OptExpression root) {
+        if (root == null) {
+            return false;
+        }
+        return isLogicalSPJ(root) || isLogicalSPJG(root);
```

Comment:
why check this logic by this way, which seems is strange.

Reply:
This is used to pre-check the mv's plan. Now only SPJ or SPJG are supported.